Add a fig-responsive execute option

cderv opened the pull request:
This is defaulting to TRUE for HTML bootstrap format only. This will mean no responsiveness for htmlwidgets in revealjs format in 
R

Adding `fig-responsive` as an execute option means that the new assignment of `htmlwidgets:::resolveSizing()` happens conditionally if the option is TRUE. Changes are done in `patch.R` and `execute.R`. 

Then `fig-responsive` default to FALSE for HTML bootstrap was trickier. Doing this meant to be able to set `format.execute["fig-responsive"] = TRUE` only for bootstrap format. In order to do that, I added a `executeFilter` following the logic of `metadataFilter` to be able to modify `execute` default option based on metadata set by the user. In our case the `theme` metadata defining is the format is Bootstrap HTML or not. I did not used `formatHasBootstrap()` because I was not sure how to use from within the format creation itself.  I did this logic for setting the default value conditionally to the theme metadata. 

@jjallaire I am opening this PR so that you can check the logic considering our previous discussion (I did not see how this would have fit in `bootstrapExtras` )

I need to tweak to not make it responsive if user has change the figure sizes.

cderv committed:
Add a fig-responsive execute option

defaulting to TRUE for HTML bootstrap format.
This will mean no responsiveness for htmlwidgets in revealjs format in R

cderv committed:
Merge branch 'main' into fig-responsive

cderv committed:
use hasBootstrapTheme()

cderv commented on src/format/html/format-html.ts:
```diff
@@ -74,6 +77,23 @@ export function htmlFormat(
   return mergeConfigs(
     createHtmlFormat(figwidth, figheight),
     {
+      execute: {
+        [kFigResponsive]: true,
+      },
```
```suggestion
```

This should not be there as we set the default in `executeFilter`

cderv committed:
remove extra execute default definition

It is set in `executeFilter`

cderv committed:
set `fig-responsive` as a metadata instead

default to true by default except for HTML presentation.
Also deactivate when minimal mode is used.

cderv commented on src/config/constants.ts:
```diff
@@ -329,6 +329,8 @@ export const kCrossrefAppendixDelim = "appendix-delim";
 export const kCrossrefChaptersAlpha = "chapters-alpha";
 export const kCrossrefChapterId = "chapter-id";
 
+export const kFigResponsive = "fig-responsive";
```
Not sure if this config should go somewhere in one of the definition interface now it is a metadata.
Currently, it is nowhere defined that this is boolean. Should it go somewhere in `types.ts` ?

jjallaire replied:
Yes, this should go in the `defaultFormat()` function. There is a bunch of other stuff you need to do for any new option -- I can check out the branch and do this once you give me the thumbs up it's okay to do

cderv replied:
Ok great. I tried to find the different places, but I must not have taken the correct other option as example. 

Or maybe because `defaultFormat()` does have yet an option for metadata. Anyway, I'll look at your change to learn

jjallaire replied:
ah, good point. we don't actually provide defaults for metadata values

cderv commented on src/format/formats-shared.ts:
```diff
@@ -96,6 +100,9 @@ export function createHtmlPresentationFormat(
   return mergeConfigs(
     createHtmlFormat(figwidth, figheight),
     {
+      metadata: {
+        [kFigResponsive]: false,
+      },
```
This should do it to set as default for all HTML presentation

cderv commented on src/format/html/format-html.ts:
```diff
@@ -75,11 +76,11 @@ export function htmlFormat(
-            ...metadata,
-            theme: "none",
-          };
+        if (!metadata[kMinimal]) {
+          return mergeConfigs(metadata, {
+            [kFigResponsive]: false,
+            theme: metadata[kTheme] === undefined ? "none" : metadata[kTheme],
+          });
```
It will now be set to TRUE by default except when minimal is used. There is no more Bootstrap conditional as discussed

cderv committed:
Do not apply responsive width if user has changed fig.width or fig.height knitr chunk option

cderv committed:
Merge branch 'main' into fig-responsive

cderv committed:
missing negation to detect change in chunk option

cderv committed:
Use explicit comparaison in if clause instead of !!

cderv commented on src/format/html/format-html.ts:
```diff
@@ -75,11 +76,11 @@ export function htmlFormat(
-            ...metadata,
-            theme: "none",
-          };
+        if (metadata[kMinimal] === true) {
```
deno linter throw me a comment if I use `!!metadata[kMinimal]` to be only `!metadata[kMinimal]`. 
I used explicit form to avoid confusion as in 
https://github.com/quarto-dev/quarto-cli/blob/217e24c7d50731cdcbeec493a5543b7a679be59b/src/format/html/format-html.ts#L395

cderv committed:
default knitr option needs to be retrieved outside of knitr to get unmodified version

cderv commented on src/resources/rmd/execute.R:
```diff
@@ -491,6 +494,42 @@ apply_slides_patch <- function(includes) {
+          # default sizing resolution
+          sizing <- htmlwidgets_resolveSizing(x, sp, standalone, knitrOptions)
+          
+          # Do not alter if user has change fig.height or fig.width option
+          if (!is.null(knitrOptions) && !identical(default_knitr_options, knitrOptions[size_options])) {
+            return(sizing)
+          }
```
This will do it per chunk. This means that if the `fig.*` options are changed on a specific chunk, it will not be modified but all the other chunk would have the transformation. 

If option is change globally `knitr::opts_chunk$set`, then all chunk would not be responsive. 

However, it is not perfect because `knitr::opts_chunk$get` will only get us the default knitr option, but not the default for the format that can be changed by Quarto. 

We could probably remove this and the two other related line above and do that later;

cderv reviewed (commented):
This is ready for the main part. 

Only the cheking of default option is not yet 100% complete.

jjallaire committed:
fig-responsive: remove attempt to preserve figure sizing

jjallaire committed:
improve handling of fig-responsive default

change metadataFilter to resolveFormat so it has access to mutating the entire format (we aren't using this now but would/will be using it if e.g. we chose to move fig-responsive to FormatRender

user can now override our default fig-responsvie behavior
